### PR TITLE
Handle non-UTF-8 OsStr on Windows

### LINKED ISSUE #339: Handle non-UTF-8 OsStr on Windows
I [asked this question on Twitter](https://twitter.com/artichokeruby/status/1191775611808731142) and got a great answer from @zkat

https://github.com/zkat/srisum-rs/blob/d375909685ae5d100dc5f832b74e09cdf16c3512/src/compute.rs#L23-L42

```rust
// NOTE: This whole dance is to allow us to print out
// the OsStr for the filenames exactly as we received
// them. It's a little bit of overkill, but it guarantees
// the user gets back what they gave and bypasses any
// weird encoding issues with these filenames.
print!("{} ", sri);

#[cfg(unix)]
let output = f.as_bytes();

#[cfg(windows)]
let output: Vec<u16> = f.encode_wide().collect();

let stdout = std::io::stdout();
let mut lock = stdout.lock();
for item in output {
    lock.write_all(&item.to_be_bytes())
        .expect("Failed to write bytes");
}
println!();
```

Our approach will be slightly different because we need a `Vec<u8>` instead of sinking to an `io::Write`.

These two functions need to be modified:

https://github.com/artichoke/artichoke/blob/17502fb91e1f6c2c0bf2462681e2d200685db746/artichoke-backend/src/fs/mod.rs#L142-L158

https://github.com/artichoke/artichoke/blob/17502fb91e1f6c2c0bf2462681e2d200685db746/artichoke-backend/src/fs/mod.rs#L171-L187

In the worst case on Windows, we'll need to allocate a new `Vec<u8>`, so change the signature of all implementations to return `Borrow<'a, [u8]>`.

Windows impl will try to convert to/from `str` first to avoid allocations and fall back to `encode_wide` + `buf.extend(wide.to_be_bytes())` in the non-UTF-8 case.

I also think as part of this change we'll want to move these functions into a platform compat module. They don't really belong in the FS.

The VM and core should only ever deal with `Vec<u8>` and convert as needed to and from `OsStr` at the edges of API boundaries.

### PR DESCRIPTION
**Pull request by dylni**
I noticed [this issue](https://github.com/artichoke/artichoke/issues/339), so I thought I would help move it forward and see how my crate handles a larger project.

Closes #339

### CODE REVIEW

**Comment by lopopolo** [comment]
Checks not running for PRs from forks is fixed in GH-574.

**Comment by dylni** [comment]
Thanks. I was wondering what was blocking them.

**Comment by lopopolo** [comment]
The crate referred to in the PR description is https://crates.io/crates/os_str_bytes:

> Traits for converting between byte sequences and platform-native strings

**Comment by lopopolo** [comment]
Thanks for the PR @dylni. I'll try to review it tonight (PDT, UTC-7).

**Comment by dylni** [comment]
Great. Thanks for the interest.

File: artichoke-backend/src/extn/core/env/backend/system.rs [hunk 1]
```diff
@@ -120,14 +117,14 @@ impl EnvType for System {
...
 
-    fn as_map(&self, interp: &Artichoke) -> Result<HashMap<Vec<u8>, Vec<u8>>, Exception> {
+    fn as_map(&self, interp: &Artichoke) -> HashMap<Vec<u8>, Vec<u8>> {
         let _ = interp;
         let mut map = HashMap::default();
         for (name, value) in std::env::vars_os() {
-            let name = ffi::os_str_to_bytes(name.as_os_str())?;
-            let value = ffi::os_str_to_bytes(value.as_os_str())?;
-            map.insert(name.to_vec(), value.to_vec());
+            let name = ffi::os_string_to_bytes(name);
```
Comment (lopopolo) [inline artichoke-backend/src/extn/core/env/backend/system.rs]: converting directly from `OsString` _into_ `Vec<u8>` saves an allocation.

File: artichoke-backend/src/extn/core/env/backend/system.rs [hunk 2]
```diff
@@ -51,12 +51,9 @@ impl EnvType for System {
...
         } else {
             let name = ffi::bytes_to_os_str(name)?;
-            if let Some(value) = std::env::var_os(name) {
-                let value = ffi::os_str_to_bytes(value.as_os_str())?;
-                Ok(Some(value.to_vec().into()))
-            } else {
-                Ok(None)
-            }
+            Ok(std::env::var_os(name)
+                .map(ffi::os_string_to_bytes)
```
Comment (lopopolo) [inline artichoke-backend/src/extn/core/env/backend/system.rs]: ditto convert from owned to owned.

**Comment by dylni** [comment]
Ok @lopopolo. `bstr` does still require valid UTF-8 on Windows, but I understand if you want to use a crate that has existed longer for this.

**Pull request closed without merging**